Thea Pallet: Set Rotation

Gauthamastro opened the pull request:
Relayers should use Queued Relayers of Thea Staking Pallet and Active Relayers of Thea pallet for operations. Thea pallet will generate an ingress message when the key updation is complete on foreign chain after which new relayers can start processing withdrawals.  Queued Relayers of Thea Staking Pallet is used to pre-generate Thea Public key and Offline Stages  for next set and update the on-chain state about it.

Gauthamastro committed:
Thea pallet logic for key rotation

Gauthamastro commented:
TODO: Sending new public key back to the on-chain state

felixfaisal commented on pallets/thea/src/lib.rs:
```diff
@@ -481,6 +510,42 @@ pub mod pallet {
+					&current_active_relayer_set
+				),
+				Error::<T>::BLSSignatureVerificationFailed
+			);
```
We should track this hash, as there exist a scenario that can lead to replay attack if relayers remain unchanged over different sets .

Gauthamastro committed:
All done.

Gauthamastro committed:
Merge branch 'Develop' into gj-session-id

felixfaisal committed:
fix: changes for thea key rotation

felixfaisal committed:
Merge branch 'gj-session-id' into fx-key-rotation

felixfaisal committed:
fix: build errors

felixfaisal committed:
fix: changes for thea key rotation (#639)

This PR introduces minor and major changes to the `Thea` to work in
accordance with `Key Rotation` Protocol

felixfaisal committed:
fix: update session length in runtime

felixfaisal committed:
fix: add index field to withdraw payload

felixfaisal committed:
fix: add index to withdrawal event

felixfaisal committed:
fix: set session time in runtime

felixfaisal committed:
fix: update session length

felixfaisal committed:
fix: session length decreased

felixfaisal committed:
fix: update event and deposit nonce

felixfaisal committed:
feat: implement key rotation  (#646)

This PR builds upon the existing PR and it has been tested e2e with
`Thea` Relayers

felixfaisal committed:
fix: add an extrinsic to reset key rotation

felixfaisal committed:
fix: check for key rotation

felixfaisal committed:
fix: changes for frontend integration

felixfaisal committed:
fix: uncomment on_initialize() hook

felixfaisal committed:
fix: set key rotation to false

felixfaisal committed:
fix: key rotation extrinsics  (#648)

This PR resolves #647

felixfaisal committed:
fix: ci checks

felixfaisal committed:
fix: comment out tests mod

felixfaisal committed:
fix: ci errors

felixfaisal committed:
fix: clippy warnings

felixfaisal committed:
refactor: apply rustfmt

felixfaisal committed:
fix: remove thea-staking tests

felixfaisal commented:
Resolves #647 
Resolves #645 (In events rather than in `struct` as mentioned in the issue)

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -468,23 +514,183 @@ pub mod pallet {
+			Ok(())
+		}
+
+		#[pallet::weight(1000)]
+		pub fn thea_relayers_reset_rotation(
```
Please add docblock as on other extrinsics

felixfaisal replied:
Done ✅

felixfaisal committed:
fix: add docblock
